src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,405 @@
+
+/**
+ * Core class for rolling back auto-update plugin failures.
+ */
+class WP_Rollback_Auto_Update {
+
+	/**
+	 * Stores handler parameters.
+	 *
+	 * @var array
+	 */
+	private $handler_args = array();
+
+	/**
+	 * Stores successfully updated plugins.
+	 *
+	 * @var array
+	 */
+	private static $processed = array();
+
+	/**
+	 * Stores fataling plugins.
+	 *
+	 * @var array
+	 */
+	private static $fatals = array();
+
+	/**
+	 * Stores `update_plugins` transient.
+	 *
+	 * @var stdClass
+	 */
+	private static $current;
+
+	/**
+	 * Stores boolean for no error from check_plugin_for_errors().
+	 *
+	 * @var bool
+	 */
+	private $no_error = false;
+
+	/**
+	 * Stores error codes.
+	 *
+	 * @var int
+	 */
+	public $error_types = E_ERROR | E_PARSE | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR;
+
+	/**
+	 * Checks the validity of the updated plugin.
+	 *
+	 * @param array|WP_Error $result     Result from WP_Upgrader::install_package().
+	 * @param array          $hook_extra Extra arguments passed to hooked filters.
+	 * @return array|WP_Error
+	 */
+	public function auto_update_check( $result, $hook_extra ) {
+		if ( is_wp_error( $result ) || ! wp_doing_cron() || ! isset( $hook_extra['plugin'] ) ) {
+			return $result;
+		}
+
+		// Already processed.
+		if ( in_array( $hook_extra['plugin'], self::$processed, true ) ) {
+			return $result;
+		}
+
+		error_log( $hook_extra['plugin'] . ' processing...' );
+
+		/**
+		 * This possibly helps to avoid a potential race condition on servers that may start to
+		 * process the next plugin for auto-updating before the handler can pick up an error from
+		 * the previously processed plugin.
+		 */
+		sleep( 2 );
+
+		$this->no_error     = false;
+		static::$current    = get_site_transient( 'update_plugins' );
+		$this->handler_args = array(
+			'handler_error' => '',
+			'result'        => $result,
+			'hook_extra'    => $hook_extra,
+		);
+
+		// Register exception and shutdown handlers.
+		$this->initialize_handlers();
+
+		self::$processed[] = $hook_extra['plugin'];
+
+		if ( is_plugin_inactive( $hook_extra['plugin'] ) ) {
+			// Working parts of plugin_sandbox_scrape().
+			wp_register_plugin_realpath( WP_PLUGIN_DIR . '/' . $hook_extra['plugin'] );
+			include WP_PLUGIN_DIR . '/' . $hook_extra['plugin'];
+		}
+
+		// Needs to run for both active and inactive plugins. Don't ask why, just accept it.
+		$this->check_plugin_for_errors( $hook_extra['plugin'] );
+		error_log( $hook_extra['plugin'] . ' auto updated.' );
+
+		return $result;
+	}
+
+	/**
+	 * Checks a new plugin version for errors.
+	 *
+	 * If an error is found, the previously installed version will be reinstalled
+	 * and an email will be sent to the site administrator.
+	 *
+	 * @param string $plugin The plugin to check.
+	 *
+	 * @throws \Exception If errors are present.
+	 *
+	 * @return void
+	 */
+	private function check_plugin_for_errors( $plugin ) {
+		global $wp_filesystem;
+
+		if ( file_exists( ABSPATH . '.maintenance' ) ) {
```

Comment:
```suggestion
		if ( $wp_filesystem->exists( ABSPATH . '.maintenance' ) ) {
```
We're using `$wp_filesystem` anyway